framework/src/eventloop/eventloop_timer.c

```diff
@@ -0,0 +1,201 @@
+/****************************************************************************
+ *
+ * Copyright 2018 Samsung Electronics All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <stdio.h>
+#include <debug.h>
+#include <libtuv/uv.h>
+#include <eventloop/eventloop.h>
+
+#include "eventloop_internal.h"
+
+typedef struct timer_cb_s {
+	timeout_callback func;
+	void *cb_data;
+} timer_cb_t;
+
+typedef struct timer_handle_s {
+	el_timer_t *handle;
+	struct timer_handle_s *next;
+} timer_handle_t;
+
+timer_handle_t *g_timers = NULL;
+
+static void add_timer_list(timer_handle_t *timer)
+{
+	timer_handle_t *iter;
+
+	if (timer == NULL) {
+		return;
+	}
+
+	if (g_timers == NULL) {
+		g_timers = timer;
+		return;
+	}
+
+	iter = g_timers;
+
+	/* Add a new timer to the end of timer list */
+	while (iter->next) {
+		iter = iter->next;
+	}
+	iter->next = timer;
+}
+
+static void remove_timer_list(el_timer_t *handle)
+{
+	timer_handle_t *iter;
+	timer_handle_t *next_iter;
+
+	if (g_timers == NULL || handle == NULL) {
+		return;
+	}
+
+	iter = g_timers;
+
+	/* Remove the first timer from the timer list */
+	if (iter->handle == handle) {
+		g_timers = iter->next;
+		EL_FREE(handle);
+		EL_FREE(iter);
+		return;
+	}
+
+	/* Search and Remove the timer from timer list */
+	while (iter->next) {
+		next_iter = iter->next;
+		if (next_iter->handle == handle) {
+			iter->next = next_iter->next;
+			EL_FREE(handle);
+			EL_FREE(next_iter);
+			break;
+		}
+		iter = next_iter;
+	}
+}
+
+static void timeout_callback_func(el_timer_t *handle)
+{
+	timer_cb_t *callback = NULL;
+
+	if (!handle || !handle->data) {
+		eldbg("Callback! timer handle is invalid\n");
+		return;
+	}
+
+	callback = handle->data;
+
+	if (callback->func) {
+		callback->func(callback->cb_data);
+	}
+
+	EL_FREE(callback);
+	EL_FREE(handle);
+}
+
+int eventloop_add_timeout(unsigned int timeout, timeout_callback func, void *data)
+{
+	timer_handle_t *timer = NULL;
+	timer_cb_t *callback = NULL;
+
+	timer = (timer_handle_t *)EL_ALLOC(sizeof(timer_handle_t));
+	if (timer == NULL) {
+		return EVENTLOOP_OUT_OF_MEMORY;
+	}
+
+	timer->handle = (el_timer_t *)EL_ALLOC(sizeof(el_timer_t));
+	if (timer->handle == NULL) {
```

Comment:
Above timer has timer_handle_t, this timer->handle has el_timer_t.
They look not matched with type and variable name.